Create harness to initialize memory from snapshot

danpoe opened the pull request:
This PR replaces PR #3755.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

danpoe committed:
Regression tests for the memory snapshot harness generator

This adds several regression tests for the feature of creating a harness that
initialises the memory from a memory snapshot.

danpoe committed:
Add const iterator to symbol table

This adds const iterators to iterate over the internal_symbols map of the symbol
table. So far, only non-const iterators existed.

danpoe committed:
Add method to get a const_targett from a location number

danpoe committed:
Add harness generator to initialise memory from a memory snapshot

This adds a harness generator to goto-harness that takes a goto program and a
memory snapshot in JSON format (via the option --memory-snapshot), and then
builds a harness that initialises the global variables from the snapshot. In
addition the harness generator provides the option --initial-location to specify
the starting point of the execution. The harness thus behaves as if the original
program would start execution at the given program location.

danpoe committed:
Output goto functions in goto-harness regression tests

Adds two invocations of cbmc --show-goto-functions to
regression/goto-harness/chain.sh to output the goto program before and after the
harness generation.

tautschnig commented on src/goto-programs/goto_program.h:
```diff
@@ -529,6 +529,23 @@ class goto_programt
     return t;
   }
 
+  optionalt<const_targett> get_target(const unsigned location_number) const
```
We have elsewhere erased the `const` qualifier from POD parameters, so it's probably better to drop it here.

hannes-steffenhagen-diffblue replied:
IMHO this is good to keep on definitions, as it still meaningful "I'm not going to modify this value in this body". I'd only remove it from declarations, where it is meaningless.

tautschnig commented on src/goto-programs/goto_program.h:
```diff
@@ -529,6 +529,23 @@ class goto_programt
+    }
+
+    return std::next(
+      instructions.begin(), location_number - start_location_number);
```
I'm not sure we provide a guarantee that location numbers are contiguous. You might want to add a `CHECK_RETURN` to make sure what you are about to return has the expected `location_number`.

xbauch replied:
Check added.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+      throw invalid_command_line_argument_exceptiont(
+        "no instruction with location number " +
+        std::to_string(*location_number) + " in function " + function);
+    }
```
I think this check should be done as part of the options parsing very early on.

danpoe replied:
At present `goto-harness` loads the goto binary only after the options checking. We could move the loading of the goto binary to before the options checking, and then give a reference to the goto model to the constructor of every harness generator. Then we could move some of the checks out of `generate()` and into `handle_option()`. Does that sound reasonable?

tautschnig replied:
Although that's a non-trivial change it would seem like a good idea to me. The earlier we know about a problem the better?

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+    split_string(initial_location, ':', start, true);
+
+    if(
+      start.size() == 0 || (start.size() >= 1 && start.front().empty()) ||
```
1) Use `start.empy()` instead of `size() == 0`
2) `start.size() >= 1` is trivially true here

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+  if(function.empty())
+  {
+    INVARIANT(
+      location_number.has_value(),
```
Should be negated (if no function is provided, then `location_number` must _not_ be set.

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+    bool_typet(),
+    id2string(function),
+    "func_init_done",
+    source_locationt(),
```
I'd use `function_symbol.location` - it's always useful to have some idea where this might belong to in terms of source code files.

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+    function_symbol.mode,
+    symbol_table);
+
+  const symbol_exprt &var = symbol.symbol_expr();
```
Don't take a reference here, `symbol_expr` returns a fresh value.

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+
+  // initialise var in __CPROVER_initialize if it is present
+  const auto f_it =
+    goto_functions.function_map.find(CPROVER_PREFIX "initialize");
```
Use the `INITIALIZE_FUNCTION` macro from `static_lifetime_init.h`

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+    "func_init_done",
+    source_locationt(),
+    function_symbol.mode,
+    symbol_table);
```
Add `symbol.value = false_exprt();` (see below)

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+  const auto f_it =
+    goto_functions.function_map.find(CPROVER_PREFIX "initialize");
+
+  if(f_it != goto_functions.function_map.end())
```
So what if this fails? I think that's where the initial value comes into play, i.e., you need to do `symbol.value = false_exprt();`

xbauch replied:
Ah, right now `validate_options` has a check for the presence of `initialize`. If absent, we throw exception. Do you want it to proceed and simply skip the modification of `initialize` altogether?

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+
+  goto_programt &goto_program = goto_function.body;
+
+  const symbolt &symbol = get_fresh_aux_symbol(
```
This might not give you a symbol of the expected lifetime. I think you may want to set `symbol.is_static_lifetime = true;` afterwards.

hannes-steffenhagen-diffblue replied:
Can you give `symbol` a more meaningful name, such as `func_init_done` or `func_init_done_symbol` if you want to keep its symbol-ness part of the name?

Also, can you add a comment here about what `func_init_done` means?

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+  const goto_programt::const_targett start_it =
+    goto_program.instructions.begin();
+
+  goto_programt::const_targett it;
```
Instead of having an uninitialised iterator you might just use `optionalt<goto_programt::const_targett> opt_it;` right here.

xbauch replied:
Fixed.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+  // modifying the goto model)
+  goto_programt::const_targett it =
+    location_number.has_value() ? goto_program.get_target(*location_number)
+                                : start_it;
```
Isn't this now done above (with very nice error handling)?

xbauch replied:
We checked that the location number exists but did not store its instruction.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+
+    code_assignt code_assign(symbol.symbol_expr(), symbol.value);
+
+    code.add(code_assign);
```
Use `std::move` or simply do `code.add(code_assignt(....));`

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+
+  for(const code_typet::parametert &parameter : code_type.parameters())
+  {
+    arguments.push_back(side_effect_expr_nondett(parameter.type()));
```
This constructor of `side_effect_expr_nondett` is deprecated, please use the one that also take as source location (just use `called_function_symbol.location`).

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+
+  code_function_callt function_call(
+    called_function_symbol.symbol_expr(), arguments);
+  code.add(function_call);
```
Use `std::move` or inline the constructor call.

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+  }
+
+  code_function_callt function_call(
+    called_function_symbol.symbol_expr(), arguments);
```
`std::move(arguments)`

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+  CHECK_RETURN(r.second);
+
+  auto f_it = goto_model.goto_functions.function_map.insert(
+    std::make_pair(function.name, goto_functiont()));
```
I'd use `emplace(function.name, goto_functiont())`

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+        "` already in "
+        "the symbol table",
+      "--" GOTO_HARNESS_GENERATOR_HARNESS_FUNCTION_NAME_OPT);
+  }
```
I'd do that check as early as possible, no need to waste time doing `add_init_section` or the likes.

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+
+  code_blockt harness_function_body;
+
+  add_assignments_to_globals(snapshot, harness_function_body);
```
Make that function return a `code_blockt` rather than passing it by reference.

xbauch replied:
Done.

kroening commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+      goto_program.insert_before(std::prev(goto_program.instructions.end()));
+
+    init_it->make_assignment(code_assignt(var, false_exprt()));
+  }
```
Now that since #4001 you can say
`goto_program.insert_before(...,  goto_programt::make_assignment(...));`
The `instructiont::make_assignment(...)` will eventually go away.

xbauch replied:
Done.

NlightNFotis commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -0,0 +1,82 @@
+
+  void insert_function(goto_modelt &goto_model, const symbolt &function) const;
+
+  std::string memory_snapshot;
```
Nit: I would change this name to `memory_snapshot_file` or something else, because its naming is slightly confusing: When I first saw it in `generate()`, I thought it had a complex data structure behind it, but it appears it's just a string indicating the file to read the json_symbol_table from.

xbauch replied:
Done.

NlightNFotis reviewed (approved):
LGTM. Just one comment.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+  const symbol_exprt &var = symbol.symbol_expr();
+
+  // initialise var in __CPROVER_initialize if it is present
+  const auto f_it =
```
Can you give this is a more meaningful name, such as `cprover_init_it` or something along those lines?

Also, can you elaborate why you need to change `__CPROVER_initialize` to make this work? (ideally as a comment in the code as well)

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+    function_symbol.mode,
+    symbol_table);
+
+  const symbol_exprt &var = symbol.symbol_expr();
```
Can you give this a more meaningful name, like `func_init_done_variable`?

xbauch replied:
Better names added.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,310 @@
+  code.add(function_call);
+}
+
+void memory_snapshot_harness_generatort::insert_function(
```
If this is supposed to insert the harness function specifically rather than just any function, can you rename this to `insert_harness_function_into_goto_model` or something like this? (I know these names are getting rather lengthy, this is just an example)

xbauch replied:
Renamed.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -0,0 +1,82 @@
+  "--memory-snapshot <file>      initialise memory from JSON memory snapshot\n"\
+  "--initial-location <func[:<n>]>\n"                                          \
+  "                              use given function and location number as "   \
+  "entry\n                              point\n"                               \
```
:pick: Looks like the formatting here got messed up a bit

xbauch replied:
Actually looks ok on the command line.

hannes-steffenhagen-diffblue commented on regression/goto-harness/chain.sh:
```diff
@@ -23,5 +23,7 @@ if [ -e "${name}-mod.gb" ] ; then
 fi
 
-$goto_harness "${name}.gb" "${name}-mod.gb" --harness-function-name $entry_point ${args} 
+$cbmc --show-goto-functions "${name}.gb"
```
Do you need this information for the tests or is this just intended to make debugging a bit easier?

xbauch replied:
Not need, so I guess debugging.

hannes-steffenhagen-diffblue reviewed (approved):
Mostly nitpicks / questions, otherwise looks good to me

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6869881bf8a0cc0de46ba6503be38caf5b205c94).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100488819

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -0,0 +1,306 @@
+
+  // initialise var in __CPROVER_initialize if it is present
+  const auto f_it =
+    goto_functions.function_map.find(CPROVER_PREFIX "initialize");
```
I'm not sure why you need this at all, shouldn't the generated harness function be enough for whatever you were going to to with `initialize`?

Commit:
Search the JSON for symbol table

In case we get an array (and not an object) we first find the right object and
then pass it to be converted.

Commit:
Fix based on comments

tautschnig commented:
@danpoe @xbauch This needs a rebase.
